lnwallet/commitment.go

```diff
@@ -0,0 +1,711 @@
+		toLocalRedeemScript,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Next, we create the script paying to the remote.
+	toRemoteScript, err := CommitScriptToRemote(
+		chanType, uint32(remoteChanCfg.CsvDelay), keyRing.RemoteKey,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Now that both output scripts have been created, we can finally create
+	// the transaction itself. We use a transaction version of 2 since CSV
+	// will fail unless the tx version is >= 2.
+	commitTx := wire.NewMsgTx(2)
+	commitTx.AddTxIn(&fundingOutput)
+
+	var (
+		localOutput  = false
+		remoteOutput = false
+	)
+
+	// Avoid creating dust outputs within the commitment transaction.
+	if amountToLocal >= localChanCfg.DustLimit {
+		commitTx.AddTxOut(&wire.TxOut{
+			PkScript: toLocalScriptHash,
+			Value:    int64(amountToLocal),
+		})
+		localOutput = true
+	}
+	if amountToRemote >= localChanCfg.DustLimit {
+		commitTx.AddTxOut(&wire.TxOut{
+			PkScript: toRemoteScript.PkScript,
+			Value:    int64(amountToRemote),
+		})
+		remoteOutput = true
+	}
+
+	// If this channel type has anchor, we'll also add those.
+	if chanType.HasAnchors() {
+		localAnchor, remoteAnchor, err := CommitScriptAnchors(
+			localChanCfg, remoteChanCfg,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// Add local anchor output only if we have a commitment output
+		// or there are HTLCs.
+		if localOutput || numHTLCs > 0 {
+			commitTx.AddTxOut(&wire.TxOut{
+				PkScript: localAnchor.PkScript,
+				Value:    int64(anchorSize),
+			})
+		}
+
+		// Add anchor output to remote only if they have a commitment
+		// output or there are HTLCs.
+		if remoteOutput || numHTLCs > 0 {
+			commitTx.AddTxOut(&wire.TxOut{
+				PkScript: remoteAnchor.PkScript,
+				Value:    int64(anchorSize),
+			})
+		}
+	}
+
+	return commitTx, nil
+}
+
+// genHtlcScript generates the proper P2WSH public key scripts for the HTLC
+// output modified by two-bits denoting if this is an incoming HTLC, and if the
+// HTLC is being applied to their commitment transaction or ours.
+func genHtlcScript(chanType channeldb.ChannelType, isIncoming, ourCommit bool,
+	timeout uint32, rHash [32]byte,
+	keyRing *CommitmentKeyRing) ([]byte, []byte, error) {
+
+	var (
+		witnessScript []byte
+		err           error
+	)
+
+	// Choose scripts based on channel type.
+	receiverHTLCScript := input.ReceiverHTLCScript
+	senderHTLCScript := input.SenderHTLCScript
+	if chanType.HasAnchors() {
+		receiverHTLCScript = input.ReceiverHTLCScriptAnchors
+		senderHTLCScript = input.SenderHTLCScriptAnchors
+	}
+
+	// Generate the proper redeem scripts for the HTLC output modified by
+	// two-bits denoting if this is an incoming HTLC, and if the HTLC is
+	// being applied to their commitment transaction or ours.
+	switch {
+	// The HTLC is paying to us, and being applied to our commitment
+	// transaction. So we need to use the receiver's version of HTLC the
+	// script.
+	case isIncoming && ourCommit:
```

Comment:
They keyring is already base on whether this is our commit our not... could we simplify this switch?

Reply:
would also need to flip the `isIncoming` based on the local/remote perspecitve. Seems to be a bigger change that would, leaving for later.